From d0a68a3b76e05bf6ca1fbd1374644e5dc9bf82e4 Mon Sep 17 00:00:00 2001
From: AlexProgrammerDE <40795980+AlexProgrammerDE@users.noreply.github.com>
Date: Sun, 27 Sep 2026 10:21:54 +0200
Subject: [PATCH] fix(vehicles): align player position when dismounting

The Bedrock player input retains the boarding position while a predicted vehicle moves. On unlink or vehicle removal, seed the player position from the vehicle before clearing the mount so the next auth input does not snap the server player back to the boarding point.
---
 .../api/model/entity/ClientPlayerEntity.java          | 11 +++++++++++
 .../protocol/packet/InteractionPackets.java           |  3 +--
 .../viabedrock/protocol/storage/EntityTracker.java    |  3 +--
 .../protocol/storage/EntityTrackerTest.java           |  4 ++++
 4 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/src/main/java/net/raphimc/viabedrock/api/model/entity/ClientPlayerEntity.java b/src/main/java/net/raphimc/viabedrock/api/model/entity/ClientPlayerEntity.java
index 870cbfaf..8e9071fd 100644
--- a/src/main/java/net/raphimc/viabedrock/api/model/entity/ClientPlayerEntity.java
+++ b/src/main/java/net/raphimc/viabedrock/api/model/entity/ClientPlayerEntity.java
@@ -416,6 +416,17 @@ public class ClientPlayerEntity extends PlayerEntity {
         this.paddlingRight = false;
     }
 
+    public void dismountVehicle(final Entity vehicle) {
+        if (this.mountRuntimeId != vehicle.runtimeId()) {
+            return;
+        }
+
+        final Position3f vehiclePosition = vehicle.position();
+        this.setPosition(vehiclePosition.add(0F, this.eyeOffset(), 0F));
+        this.setMountEntityRuntimeId(-1);
+        this.requestedDismount = false;
+    }
+
     @Override
     public void setMountEntityRuntimeId(final long runtimeId) {
         if (this.mountRuntimeId != runtimeId) {
diff --git a/src/main/java/net/raphimc/viabedrock/protocol/packet/InteractionPackets.java b/src/main/java/net/raphimc/viabedrock/protocol/packet/InteractionPackets.java
index 4a69f711..d72086ba 100644
--- a/src/main/java/net/raphimc/viabedrock/protocol/packet/InteractionPackets.java
+++ b/src/main/java/net/raphimc/viabedrock/protocol/packet/InteractionPackets.java
@@ -376,8 +376,7 @@ public final class InteractionPackets {
 
                     if (linkType.toEntityUniqueId() == entityTracker.getClientPlayer().uniqueId()) { // TODO: This could be applied to all passengers not just players
                         // The player is no longer riding an entity, update the state
-                        entityTracker.getClientPlayer().setMountEntityRuntimeId(-1);
-                        entityTracker.getClientPlayer().setRequestedDismount(false);
+                        entityTracker.getClientPlayer().dismountVehicle(vehicle);
                     }
                 }
             }
diff --git a/src/main/java/net/raphimc/viabedrock/protocol/storage/EntityTracker.java b/src/main/java/net/raphimc/viabedrock/protocol/storage/EntityTracker.java
index 221b3c7d..13ae3128 100644
--- a/src/main/java/net/raphimc/viabedrock/protocol/storage/EntityTracker.java
+++ b/src/main/java/net/raphimc/viabedrock/protocol/storage/EntityTracker.java
@@ -124,8 +124,7 @@ public class EntityTracker extends StoredObject {
             trackedEntity.removePassenger(entity.uniqueId());
         }
         if (this.clientPlayerEntity != null && this.clientPlayerEntity.mountEntityRuntimeId() == entity.runtimeId()) {
-            this.clientPlayerEntity.setMountEntityRuntimeId(-1);
-            this.clientPlayerEntity.setRequestedDismount(false);
+            this.clientPlayerEntity.dismountVehicle(entity);
         }
         entity.remove();
     }
diff --git a/src/test/java/net/raphimc/viabedrock/protocol/storage/EntityTrackerTest.java b/src/test/java/net/raphimc/viabedrock/protocol/storage/EntityTrackerTest.java
index 23575461..c76e603f 100644
--- a/src/test/java/net/raphimc/viabedrock/protocol/storage/EntityTrackerTest.java
+++ b/src/test/java/net/raphimc/viabedrock/protocol/storage/EntityTrackerTest.java
@@ -14,6 +14,7 @@ import com.viaversion.viaversion.api.minecraft.entities.EntityTypes26_3;
 import net.raphimc.viabedrock.api.model.entity.ClientPlayerEntity;
 import net.raphimc.viabedrock.api.model.entity.Entity;
 import net.raphimc.viabedrock.protocol.model.PlayerAbilities;
+import net.raphimc.viabedrock.protocol.model.Position3f;
 import org.junit.jupiter.api.Test;
 
 import java.lang.reflect.Proxy;
@@ -60,12 +61,15 @@ class EntityTrackerTest {
         final ClientPlayerEntity player = new ClientPlayerEntity(user, 1, UUID.randomUUID(), new PlayerAbilities(1, (byte) 0, (byte) 0));
         tracker.addEntity(player, false);
         final Entity vehicle = tracker.addEntity(10, 100, EntityTypes26_3.PIG.identifier(), EntityTypes26_3.PIG);
+        vehicle.setPosition(new Position3f(12F, 64F, 8F));
+        player.setPosition(new Position3f(0F, 65.62F, 0F));
         player.setMountEntityRuntimeId(vehicle.runtimeId());
         vehicle.addPassenger(player.uniqueId());
 
         tracker.removeEntity(vehicle);
 
         assertEquals(-1, player.mountEntityRuntimeId());
+        assertEquals(new Position3f(12F, 65.62F, 8F), player.position());
         assertSame(player, tracker.getEntityByUid(player.uniqueId()));
     }
 
-- 
2.55.0


